README.md
[![Version](https://img.shields.io/badge/version-0.2.0-4B9081.svg)](https://github.com/steevanb/php-version-comparator/tree/0.2.0)
[![License](https://poser.pugx.org/steevanb/version-comparator/license)](https://github.com/steevanb/php-version-comparator/blob/master/LICENSE)
[![PHP](https://img.shields.io/badge/php-^5.3||^7.0||^8.0-blue.svg)](https://php.net)
![Lines](https://img.shields.io/badge/code%20lines-4,561-blue.svg)
![Downloads](https://poser.pugx.org/steevanb/version-comparator/downloads)

[![CI](https://github.com/steevanb/php-version-comparator/actions/workflows/ci.yml/badge.svg?branch=master)](https://github.com/steevanb/php-version-comparator/actions/workflows/ci.yml)
![Tests](https://img.shields.io/badge/tests-40-blue.svg)
![Assertions](https://img.shields.io/badge/assertions-70-blue.svg)
![Coverage](https://img.shields.io/badge/coverage-70%25-success.svg)
![Infection](https://img.shields.io/badge/infection-95%25-success.svg)

## steevanb/version-comparator

Add classes to compare versions easier than with `compare_version()` and `PHP_VERSION`.

[Changelog](changelog.md)

## Installation

```
composer require steevanb/version-comparator ^0.2
```

## Compare with current PHP version

You can compare current PHP version with 2 versions to know if current php version is between this 2 versions:

```php
PhpVersionComparator::isBetween('8', '9'); // return true is PHP is >= 8.0.0 and < 9.0.0
PhpVersionComparator::isBetween('8.0', '8.1'); // return true is PHP is >= 8.0.0 and < 8.1.0
PhpVersionComparator::isBetween('8.0.0', '8.0.2'); // return true is PHP is >= 8.0.0 and < 8.0.2
```

## Shortcuts for each PHP major and minor version

`PhpVersionComparator` have a shortcut who call `isBetween()` for each major and minor version:

```php
// For PHP 5.3 to 5.6 
PhpVersionComparator::isPhp5(); // return true is PHP is >= 5.0.0 and < 6.0.0
PhpVersionComparator::isPhp53(); // return true is PHP is >= 5.3.0 and < 5.4.0
PhpVersionComparator::isPhp54(); // return true is PHP is >= 5.4.0 and < 5.5.0
PhpVersionComparator::isPhp55(); // return true is PHP is >= 5.5.0 and < 5.6.0
PhpVersionComparator::isPhp56(); // return true is PHP is >= 5.6.0 and < 5.7.0

// For PHP 7.0 to 7.4
PhpVersionComparator::isPhp7(); // return true is PHP is >= 7.0.0 and < 8.0.0
PhpVersionComparator::isPhp70(); // return true is PHP is >= 7.0.0 and < 7.1.0
PhpVersionComparator::isPhp71(); // return true is PHP is >= 7.1.0 and < 7.2.0
PhpVersionComparator::isPhp72(); // return true is PHP is >= 7.2.0 and < 7.3.0
PhpVersionComparator::isPhp73(); // return true is PHP is >= 7.3.0 and < 7.4.0
PhpVersionComparator::isPhp74(); // return true is PHP is >= 7.4.0 and < 7.5.0

// For PHP 8.0 to latest 8.x
PhpVersionComparator::isPhp8(); // return true is PHP is >= 8.0.0 and < 9.0.0
PhpVersionComparator::isPhp80(); // return true is PHP is >= 8.0.0 and < 8.1.0
PhpVersionComparator::isPhp81(); // return true is PHP is >= 8.1.0 and < 8.2.0
```

## Exception instead of returning the result

If you want to throw an exception instead of returning the result,
each method have it's `assert()` version who throw a `VersionIsNotBetweenException` exception:

```php
PhpVersionComparator::assertIsBetween('8', '9');
PhpVersionComparator::assertIsPhp5();
// etc
```


changelog.md
### master

### [0.2.0](../../compare/0.1.0...0.2.0) - 2021-10-23

- Rename namespace `Steevanb\PhpVersionCompare\PhpVersion` to `Steevanb\VersionComparator\Comparator`
- Rename project to `steevanb/version-comparator`

### [0.1.0](../../compare/0.0.0...0.1.0) - 2021-10-23

- Rename `PhpVersionCompare` to `PhpVersionComparator` and move to `Steevanb\PhpVersionCompare\PhpVersion`
- Add `assert()` methods for each major and minor PHP version

### 0.0.0 - 2021-10-23

- Add `PhpVersionCompare`


tests/Php56/PhpVersionCompare/Php5Test.php
<?php

namespace Steevanb\PhpVersionCompare\Tests\Php56\PhpVersionCompare;

use PHPUnit\Framework\TestCase;
use Steevanb\PhpVersionCompare\PhpVersionCompare;

class Php5Test extends TestCase
{
    public function testIsPhp5()
    {
        static::assertTrue(PhpVersionCompare::isPhp5());
    }

    public function testIsPhp53()
    {
        static::assertFalse(PhpVersionCompare::isPhp53());
    }

    public function testIsPhp54()
    {
        static::assertFalse(PhpVersionCompare::isPhp54());
    }

    public function testIsPhp55()
    {
        static::assertFalse(PhpVersionCompare::isPhp55());
    }

    public function testIsPhp56()
    {
        static::assertTrue(PhpVersionCompare::isPhp56());
    }
}


tests/Php56/PhpVersionCompare/Php7Test.php
<?php

namespace Steevanb\PhpVersionCompare\Tests\Php56\PhpVersionCompare;

use PHPUnit\Framework\TestCase;
use Steevanb\PhpVersionCompare\PhpVersionCompare;

final class Php7Test extends TestCase
{
    public function testIsPhp7()
    {
        static::assertFalse(PhpVersionCompare::isPhp7());
    }

    public function testIsPhp70()
    {
        static::assertFalse(PhpVersionCompare::isPhp70());
    }

    public function testIsPhp71()
    {
        static::assertFalse(PhpVersionCompare::isPhp71());
    }

    public function testIsPhp72()
    {
        static::assertFalse(PhpVersionCompare::isPhp72());
    }

    public function testIsPhp73()
    {
        static::assertFalse(PhpVersionCompare::isPhp73());
    }

    public function testIsPhp74()
    {
        static::assertFalse(PhpVersionCompare::isPhp74());
    }
}


tests/Php56/PhpVersionCompare/Php8Test.php
<?php

namespace Steevanb\PhpVersionCompare\Tests\Php56\PhpVersionCompare;

use PHPUnit\Framework\TestCase;
use Steevanb\PhpVersionCompare\PhpVersionCompare;

final class Php8Test extends TestCase
{
    public function testIsPhp8()
    {
        static::assertFalse(PhpVersionCompare::isPhp8());
    }

    public function testIsPhp80()
    {
        static::assertFalse(PhpVersionCompare::isPhp80());
    }

    public function testIsPhp81()
    {
        static::assertFalse(PhpVersionCompare::isPhp81());
    }
}


src/Comparator/PhpVersionComparator.php
<?php

namespace Steevanb\VersionComparator\Comparator;

use Steevanb\VersionComparator\Exception\VersionIsNotBetweenException;

class PhpVersionComparator
{
    const VERSION_5 = '5.0.0';
    const VERSION_5_3 = '5.3.0';
    const VERSION_5_4 = '5.4.0';
    const VERSION_5_5 = '5.5.0';
    const VERSION_5_6 = '5.6.0';

    const VERSION_7 = '7.0.0';
    const VERSION_7_0 = '7.0.0';
    const VERSION_7_1 = '7.1.0';
    const VERSION_7_2 = '7.2.0';
    const VERSION_7_3 = '7.3.0';
    const VERSION_7_4 = '7.4.0';

    const VERSION_8 = '8.0.0';
    const VERSION_8_0 = '8.0.0';
    const VERSION_8_1 = '8.1.0';

    /** @return bool */
    public static function isPhp5()
    {
        return static::isBetween(static::VERSION_5, '6.0.0');
    }

    public static function assertIsPhp5()
    {
        static::assertIsBetween(static::VERSION_5, '6.0.0');
    }

    /** @return bool */
    public static function isPhp53()
    {
        return static::isBetween(static::VERSION_5_3, static::VERSION_5_4);
    }

    /** @return void */
    public static function assertIsPhp53()
    {
        static::assertIsBetween(static::VERSION_5_3, static::VERSION_5_4);
    }

    /** @return bool */
    public static function isPhp54()
    {
        return static::isBetween(static::VERSION_5_4, static::VERSION_5_5);
    }

    /** @return void */
    public static function assertIsPhp54()
    {
        static::assertIsBetween(static::VERSION_5_4, static::VERSION_5_5);
    }

    /** @return bool */
    public static function isPhp55()
    {
        return static::isBetween(static::VERSION_5_5, static::VERSION_5_6);
    }

    /** @return void */
    public static function assertIsPhp55()
    {
        static::assertIsBetween(static::VERSION_5_5, static::VERSION_5_6);
    }

    /** @return bool */
    public static function isPhp56()
    {
        return static::isBetween(static::VERSION_5_6, '5.7.0');
    }

    /** @return void */
    public static function assertIsPhp56()
    {
        static::assertIsBetween(static::VERSION_5_6, '5.7.0');
    }

    /** @return bool */
    public static function isPhp7()
    {
        return static::isBetween(static::VERSION_7, static::VERSION_8);
    }

    /** @return void */
    public static function assertIsPhp7()
    {
        static::assertIsBetween(static::VERSION_7, static::VERSION_8);
    }

    /** @return bool */
    public static function isPhp70()
    {
        return static::isBetween(static::VERSION_7_0, static::VERSION_7_1);
    }

    /** @return void */
    public static function assertIsPhp70()
    {
        static::assertIsBetween(static::VERSION_7_0, static::VERSION_7_1);
    }

    /** @return bool */
    public static function isPhp71()
    {
        return static::isBetween(static::VERSION_7_1, static::VERSION_7_2);
    }

    /** @return void */
    public static function assertIsPhp71()
    {
        static::assertIsBetween(static::VERSION_7_1, static::VERSION_7_2);
    }

    /** @return bool */
    public static function isPhp72()
    {
        return static::isBetween(static::VERSION_7_2, static::VERSION_7_3);
    }

    /** @return void */
    public static function assertIsPhp72()
    {
        static::assertIsBetween(static::VERSION_7_2, static::VERSION_7_3);
    }

    /** @return bool */
    public static function isPhp73()
    {
        return static::isBetween(static::VERSION_7_3, static::VERSION_7_4);
    }

    /** @return void */
    public static function assertIsPhp73()
    {
        static::assertIsBetween(static::VERSION_7_3, static::VERSION_7_4);
    }

    /** @return bool */
    public static function isPhp74()
    {
        return static::isBetween(static::VERSION_7_4, '7.5.0');
    }

    /** @return void */
    public static function assertIsPhp74()
    {
        static::assertIsBetween(static::VERSION_7_4, '7.5.0');
    }

    /** @return bool */
    public static function isPhp8()
    {
        return static::isBetween(static::VERSION_8, '9.0.0');
    }

    /** @return void */
    public static function assertIsPhp8()
    {
        static::assertIsBetween(static::VERSION_8, '9.0.0');
    }

    /** @return bool */
    public static function isPhp80()
    {
        return static::isBetween(static::VERSION_8_0, static::VERSION_8_1);
    }

    /** @return void */
    public static function assertIsPhp80()
    {
        static::assertIsBetween(static::VERSION_8_0, static::VERSION_8_1);
    }

    /** @return bool */
    public static function isPhp81()
    {
        return static::isBetween(static::VERSION_8_1, '8.2.0');
    }

    /** @return void */
    public static function assertIsPhp81()
    {
        static::assertIsBetween(static::VERSION_8_1, '8.2.0');
    }

    /**
     * @param string $version1
     * @param string $version2
     * @return bool
     */
    public static function isBetween($version1, $version2)
    {
        return
            version_compare(PHP_VERSION, $version1, '>=')
            && version_compare(PHP_VERSION, $version2, '<');
    }

    /**
     * @param string $version1
     * @param string $version2
     * @return void
     */
    public static function assertIsBetween($version1, $version2)
    {
        if (static::isBetween($version1, $version2) === false) {
            throw new VersionIsNotBetweenException(PHP_VERSION, $version1, $version2);
        }
    }
}


src/PhpVersionCompare.php
<?php

namespace Steevanb\PhpVersionCompare;

class PhpVersionCompare
{
    const VERSION_5 = '5';
    const VERSION_5_3 = '5.3';
    const VERSION_5_4 = '5.4';
    const VERSION_5_5 = '5.5';
    const VERSION_5_6 = '5.6';

    const VERSION_7 = '7';
    const VERSION_7_0 = '7.0';
    const VERSION_7_1 = '7.1';
    const VERSION_7_2 = '7.2';
    const VERSION_7_3 = '7.3';
    const VERSION_7_4 = '7.4';

    const VERSION_8 = '8';
    const VERSION_8_0 = '8.';
    const VERSION_8_1 = '8.1';

    /** @return bool */
    public static function isPhp5()
    {
        return static::isBetween(static::VERSION_5, '6');
    }

    /** @return bool */
    public static function isPhp53()
    {
        return static::isBetween(static::VERSION_5_3, static::VERSION_5_4);
    }

    /** @return bool */
    public static function isPhp54()
    {
        return static::isBetween(static::VERSION_5_4, static::VERSION_5_5);
    }

    /** @return bool */
    public static function isPhp55()
    {
        return static::isBetween(static::VERSION_5_5, static::VERSION_5_6);
    }

    /** @return bool */
    public static function isPhp56()
    {
        return static::isBetween(static::VERSION_5_6, '5.7');
    }

    /** @return bool */
    public static function isPhp7()
    {
        return static::isBetween(static::VERSION_7, static::VERSION_8);
    }

    /** @return bool */
    public static function isPhp70()
    {
        return static::isBetween(static::VERSION_7_0, static::VERSION_7_1);
    }

    /** @return bool */
    public static function isPhp71()
    {
        return static::isBetween(static::VERSION_7_1, static::VERSION_7_2);
    }

    /** @return bool */
    public static function isPhp72()
    {
        return static::isBetween(static::VERSION_7_2, static::VERSION_7_3);
    }

    /** @return bool */
    public static function isPhp73()
    {
        return static::isBetween(static::VERSION_7_3, static::VERSION_7_4);
    }

    /** @return bool */
    public static function isPhp74()
    {
        return static::isBetween(static::VERSION_7_4, '7.5');
    }

    /** @return bool */
    public static function isPhp8()
    {
        return static::isBetween(static::VERSION_8, '9');
    }

    /** @return bool */
    public static function isPhp80()
    {
        return static::isBetween(static::VERSION_8_0, static::VERSION_8_1);
    }

    /** @return bool */
    public static function isPhp81()
    {
        return static::isBetween(static::VERSION_8_1, '8.2');
    }

    /**
     * @param string $version1
     * @param string $version2
     * @return bool
     */
    public static function isBetween($version1, $version2)
    {
        return
            version_compare(PHP_VERSION, $version1, '>=')
            && version_compare(PHP_VERSION, $version2, '<');
    }
}
